pkg/server/singleprocess/service_pipeline.go

```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
 	}); err != nil {
 		return nil, err
 	}
-	run, err := s.state(ctx).PipelineRunGetLatest(pipeline.Id)
+
+	pipelineRun, err := s.state(ctx).PipelineRunGetLatest(pipeline.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build out all of the queued job requests for running this pipelines steps
+	stepJobs, pipelineRun, stepIds, err := s.buildStepJobs(ctx, log, req,
+		make(map[string]interface{}), nodeIdMap, make(map[string][]string), pipeline, pipelineRun)
 	if err != nil {
 		return nil, err
 	}
 
-	// Generate the jobs for each of the steps
+	// Get the ordered jobs.
+	var jobIds []string
+	jobMap := map[string]*pb.Ref_PipelineStep{}
+	for _, v := range stepGraph.KahnSort() {
+		// Look up step name and ref by the assigned node ID from graph generation
+		nodeId := v.(string)
+		stepRef, ok := nodeIdMap[nodeId]
+		if !ok {
+			return nil, status.Errorf(codes.Internal,
+				"could not get pipeline step ref for node id %q", nodeId)
+		} else if stepRef == nil {
+			return nil, status.Errorf(codes.Internal,
+				"node id %q returned a nil pipeline step ref", nodeId)
+		}
+
+		// get the generated queued job request
+		jobId, ok := stepIds[nodeId]
+		if !ok {
+			// NOTE(briancain):
+			// This could be better. It's basically here because we want to keep track
+			// of an embedded pipeline step ref within the step graph, but it doesn't have
+			// a "job id" because the root of the embedded pipeline is the actual ID where
+			// this is simply a reference.
+			// We don't want to add it as a job id because it doesn't actually create a job.
+			// jobId = "embedded-pipeline-ref-" + nodeId
+			continue
+		}
+
+		jobIds = append(jobIds, jobId)
+		stepRef.RunSequence = pipelineRun.Sequence
+		jobMap[jobId] = stepRef
+	}
+
+	pipelineRun.State = pb.PipelineRun_STARTING
+	if err = s.state(ctx).PipelineRunPut(pipelineRun); err != nil {
+		return nil, err
+	}
+
+	// Queue all the jobs atomically
+	if _, err := s.queueJobMulti(ctx, stepJobs); err != nil {
+		return nil, err
+	}
+
+	return &pb.RunPipelineResponse{
+		JobId:     jobIds[0],
+		AllJobIds: jobIds,
+		JobMap:    jobMap,
+	}, nil
+}
+
+func (s *Service) buildStepJobs(
+	ctx context.Context,
+	log hclog.Logger,
+	req *pb.RunPipelineRequest,
+	visitedPipelines map[string]interface{},
+	nodeIdMap map[string]*pb.Ref_PipelineStep,
+	parentDep map[string][]string,
+	pipeline *pb.Pipeline,
+	pipelineRun *pb.PipelineRun,
+) ([]*pb.QueueJobRequest, *pb.PipelineRun, map[string]string, error) {
+	if len(visitedPipelines) != 0 {
+		// Determine if we've already visisted this pipeline and included its jobs.
+		// Otherwise we'll get stuck in a cycle. This only really works because
+		// pipeline names are unique for a project. If we ever start allowing for
+		// pipelines across projects we'll need to namespace this value.
```

Comment:
```suggestion
		// Determine if we've already visited this pipeline and included its jobs.
		// Otherwise, we'll get stuck in a cycle. This only works because
		// pipeline names are unique for a project. If we ever allow
		// pipelines across projects, we'll need to namespace this value.
```